src/datasources/email/email.datasource.ts

```diff
@@ -0,0 +1,157 @@
+import { Inject, Injectable } from '@nestjs/common';
+import * as postgres from 'postgres';
+import { IEmailDataSource } from '@/domain/interfaces/email.datasource.interface';
+import { EmailVerificationCode } from '@/domain/email/entities/email-verification-code.entity';
+import { Email } from '@/datasources/email/entities/email.entity';
+import { VerificationStatus } from '@/datasources/email/entities/verification.status';
+import { EmailAddressDoesNotExistError } from '@/datasources/email/errors/email-address-does-not-exist.error';
+import { EmailAlreadyVerifiedError } from '@/datasources/email/errors/email-already-verified.error';
+import { InvalidVerificationCodeError } from '@/datasources/email/errors/invalid-verification-code.error';
+import { ExpiredVerificationCodeError } from '@/datasources/email/errors/expired-verification-code.error';
+import { IConfigurationService } from '@/config/configuration.service.interface';
+
+@Injectable()
+export class EmailDatasource implements IEmailDataSource {
+  private readonly verificationExpireTimeInSeconds: number;
+
+  constructor(
+    @Inject('DB_INSTANCE') private readonly sql: postgres.Sql,
+    @Inject(IConfigurationService) configurationService: IConfigurationService,
+  ) {
+    this.verificationExpireTimeInSeconds = configurationService.getOrThrow(
+      'email.verificationExpireTimeInSeconds',
+    );
+  }
+
+  private _resetVerificationCode(
+    sql: postgres.Sql,
+    emailId: number,
+  ): postgres.PendingQuery<VerificationStatus[]> {
+    // TODO proper code generation
+    const code = Math.floor(Math.random() * 100_000);
```

Comment:
Out of scope, but we need to consider how the email provider handles verification as well (if they do).

Reply:
I agree. However I saw the provider as one to distribute emails only and not for having some kind of registration with the user. This is something that should be explored when the integration with the provider is done @hectorgomezv